### DISCUSSION TITLE
Why is altitude error used for vertical speed correction?

### DISCUSSION THREAD

**Discussion by avsaase** [iNavFlight/inav#6606] [Q&A] [answered]
In `estimationCalculateCorrection_Z()`, when having a valid baro reading, the difference between the estimated altitude from `estimationPredict()` and the baro is used to correct the altitude estimation. That makes sense. But for the vertical speed estimation, the correction is also based on this altitude residual, with the square of the position weight, instead of using some sort of baro vertical speed measure. I don't understand why that is.

https://github.com/iNavFlight/inav/blob/1356480ebb8f37daec321f365d067e7a42fe1d81/src/main/navigation/navigation_pos_estimator.c#L563-L566

The same applies to the GPS when you don't have a baro and are using `use_gps_no_baro`, but here there is a double correction using both the altitude residual and the residual from the GPS vertical speed. 

https://github.com/iNavFlight/inav/blob/1356480ebb8f37daec321f365d067e7a42fe1d81/src/main/navigation/navigation_pos_estimator.c#L597-L599

And a more general question, why are the altitude and vertical speed estimation are so decoupled from each other? Wouldn't it make more sense to have one estimation for altitude and take the difference with the previous value for the vertical speed? This would avoid inconsistencies where the altitude is increasing but the vario indicates negative vertical speed and vice versa.

Thanks.

**Comment by DzikuVx** [comment]
@digitalentity ?

**Comment by digitalentity** [comment, marked as answer]
This is WAI and based on Alpha-Beta filter math: https://en.wikipedia.org/wiki/Alpha_beta_filter

The filter provides both a coordinate and speed consuming a single observation (coordinate). We add an additional observation (acceleration) to improve response times for disturbances, but it has a drawback of causing the filter to disconverge occasionally (i.e. increasing coordinate, but negative speed). However, the Alpha-Beta filter if numerically stabile (unlike Kalman) and will converge again so this is deemed to be safe.

**Reply by avsaase** [reply]
Thanks, this explains a lot. But isn't there a scaling issue, with both the altitude and vertical speed being corrected with `* ctx->dt`?

EDIT: actually not necessarily a scaling issue, the velocity correction is just very small. `beta = alpha^2 * dT` to be exact. With the default `w_z_baro_p = 0.35` and a 2kHz loop frequency this means `beta = 6E-5`. Here's my math: [alpha-beta-filter.pdf](https://github.com/iNavFlight/inav/files/5988776/alpha-beta-filter.pdf).

**Reply by digitalentity** [reply]
Both coordinate and velocity correction is scaled by dT to make weights independent of the update rate. We can apply correction every 1 second with full weight, every 0.5 seconds with 1/2 weight and so on. Otherwise we'll have to adjust the weights every time we change the update rate (this is how it was back in the early days when INAV was just a PR to Cleanflight :laughing:)

**Reply by digitalentity** [reply]
But you're probably right about lack of ∆T for `beta` :thinking: 
Feel free to experiment with that and see if that improves velocity tracking.

**Reply by avsaase** [reply]
I will experiment with 47b84d64fd78fd8448bf000ad80bda3983824675. I just hope my quad doesn't fly away because of this 😅

**Reply by avsaase** [reply]
Okay, that was a disaster. As soon as I switched into POS HOLD the quad tried to take off with pulses of 100% throttle. Looking at the log I see the vertical speed is all over the place. Makes sense since I increased the correction by a factor of 2000. A well, you live and learn.

![altitudeplot](https://user-images.githubusercontent.com/880421/108098916-3e691480-7084-11eb-8b04-fa6a34564fad.png)

**Comment by DzikuVx** [comment]
@avsaase on a off-topic note, could you drop me an email? You should find my address and I can't find yours :)

**Reply by avsaase** [reply]
Done. I used the address on your GitHub profile.
